supernode/services/userdataprocess/task.go

```diff
@@ -0,0 +1,528 @@
+		}, errors.Errorf("error on action: %w", actionErr)
+	}
+
+	// only primary node start this action
+	var processResult userdata.ProcessResult
+	if task.ConnectedTo == nil {
+		<-task.NewAction(func(ctx context.Context) error {
+			log.WithContext(ctx).Debug("waiting for signature from peers")
+			for {
+				select {
+				case <-ctx.Done():
+					err := ctx.Err()
+					if err != nil {
+						actionErr = errors.Errorf("waiting for SuperNodeRequest from peers cancelled or timeout: %w", err)
+					}
+					return nil
+				case <-task.allPeersSNDatasReceived:
+					log.WithContext(ctx).Debugf("all SuperNodeRequest received so start validation")
+					resp, err := task.verifyPeersUserdata(ctx)
+					if err != nil {
+						actionErr = errors.Errorf("fail to verifyPeersUserdata: %w", err)
+					}
+					processResult = resp
+					return nil
+				}
+			}
+		})
+	}
+
+	return processResult, actionErr
+}
+
+// ReceiveUserdata get the userdata from database
+func (task *Task) ReceiveUserdata(ctx context.Context, userpastelid string) (userdata.ProcessRequest, error) {
+	log.WithContext(ctx).Debugf("ReceiveUserdata on user PastelID: %s", userpastelid)
+
+	// only primary node start this action
+	return task.Service.databaseOps.ReadUserData(ctx, userpastelid)
+}
+
+// Sign and send SNDataSigned if not primary
+func (task *Task) signAndSendSNDataSigned(ctx context.Context, sndata userdata.SuperNodeRequest, isPrimary bool) error {
+	log.WithContext(ctx).Debugf("signAndSendSNDataSigned begin to sign SuperNodeRequest")
+	signature, err := task.pastelClient.Sign(ctx, []byte(sndata.UserdataHash+sndata.UserdataResultHash), task.config.PastelID, task.config.PassPhrase, "ed448")
+	if err != nil {
+		return errors.Errorf("failed to sign sndata %w", err)
+	}
+	if !isPrimary {
+		sndata.HashSignature = hex.EncodeToString(signature)
+		sndata.NodeID = task.config.PastelID
+		log.WithContext(ctx).Debug("send signed sndata to primary node")
+		if _, err := task.ConnectedTo.ProcessUserdata.SendUserdataToPrimary(ctx, sndata); err != nil {
+			return errors.Errorf("failed to send signature to primary node %s at address %s %w", task.ConnectedTo.ID, task.ConnectedTo.Address, err)
+		}
+	}
+	return nil
+}
+
+// AddPeerSNDataSigned gather all the Userdata with signature from other Supernodes
+func (task *Task) AddPeerSNDataSigned(ctx context.Context, snrequest userdata.SuperNodeRequest) error {
+	log.WithContext(ctx).Debugf("addPeerSNDataSigned begin to aggregate SuperNodeRequest")
+
+	task.peersSNDataSignedMtx.Lock()
+	defer task.peersSNDataSignedMtx.Unlock()
+
+	var actionErr error
+	<-task.NewAction(func(ctx context.Context) error {
+		log.WithContext(ctx).Debugf("receive supernode userdata result signed from node %s", snrequest.NodeID)
+		if node := task.accepted.ByID(snrequest.NodeID); node == nil {
+			actionErr = errors.Errorf("node %s not in accepted list", snrequest.NodeID)
+			return nil
+		}
+
+		task.peersSNDataSigned[snrequest.NodeID] = snrequest
```

Comment:
locked ?

Reply:
It's not thread safe so it's locked on purpose. The flow is exactly similar to AddPeerArticketSignature in art reg flow.

Reply:
we locked it already
thank you